src/DNSC/Cache.hs

```diff
@@ -172,23 +172,28 @@ insertRRs now rrs rank c = insertRRSet =<< takeRRSet rrs
 @
  -}
 insert :: Timestamp -> Key -> TTL -> CRSet -> Ranking -> Cache -> Maybe Cache
-insert now k@(Key dom typ cls) ttl crs rank c =
-  maybe inserted withOldRank lookupRank
+insert now k@(Key dom typ cls) ttl crs rank cache@(Cache c xsz) =
+  maybe sized withOldRank lookupRank
   where
     lookupRank =
       lookup_ now (\_ _ (Val _ r) -> r)
-      dom typ cls c
+      dom typ cls cache
     withOldRank r = do
       guard $ rank > r
-      inserted
+      sized
     eol = now <+ ttl
-    inserted =
-      return $ PSQ.insert k eol (Val crs rank) c
+    sized
+      | PSQ.size inserted <= xsz  =  return $ Cache inserted xsz
+      | otherwise                 =  do
+          (_, _, _, deleted) <- PSQ.minView inserted
```

Comment:
I don't like insert-then-delete. Rather, I like delete-then-insert.

Reply:
updated to delete-then-insert.
https://github.com/khibino/dns-resolver/pull/14/commits/3bb15c9a3a9559d72a8a2a8a0d9a5df5b32ef5f1